ibridges/meta.py

```diff
@@ -142,6 +142,36 @@ def __getitem__(self, key: Union[str, Sequence[str]]) -> MetaDataItem:
                              "units as well, for example: meta[key, value, units].")
         return all_items[0]
 
+    def __setitem__(self, key: Union[str, Sequence[Union[str, None]]], other: Sequence[str]):
+        """Set metadata items like a dictionary of tuples.
+
+        Parameters
+        ----------
+        key
+            The key to get the metadata for.
+        other
+            Key, value, units to set the metadata item to. Units is optional.
+
+        Raises
+        ------
+        TypeError:
+            If the other parameter is a string.
+        ValueError:
+            If the item already exists.
+
+        Examples
+        --------
+        >>> meta["key"] = ("key", "new_value", "new_units")
+        >>> meta["key"] = ("new_key", "old_value")
+
+        """
+        if isinstance(other, str):
+            raise TypeError("Cannot set the metadata item to a single string value. "
+                            f"Use meta[{key}].key = \"{other}\" to change only the key "
+                            "for example.")
+        self[key].update(*other)
+
+
     def add(self, key: str, value: str, units: Optional[str] = None):
```

Comment:
The error message if `key=None`  or `value = None` is not so user-friendly:

```
irods_path.meta.add('key1', None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[27], line 1
----> 1 irods_path.meta.add('key1', None)

File ~/git-repos/iBridges/ibridges/meta.py:212, in MetaData.add(self, key, value, units)
    210         if re.match(self.blacklist, key):
    211             raise ValueError(f"ADD META: Key must not start with {self.blacklist}.")
--> 212     self.item.metadata.add(key, value, units)
    213 except irods.exception.CAT_NO_ACCESS_PERMISSION as error:
    214     raise PermissionError("UPDATE META: no permissions") from error

File /usr/local/lib/python3.12/site-packages/irods/meta.py:143, in iRODSMetaCollection.add(self, *args, **opts)
    139 """
    140 Add as iRODSMeta to a key
    141 """
    142 meta = self._get_meta(*args)
--> 143 self._manager.add(self._model_cls, self._path, meta, **opts)
    144 self._reset_metadata()

File /usr/local/lib/python3.12/site-packages/irods/manager/metadata_manager.py:99, in MetadataManager.add(self, model_cls, path, meta, **opts)
     97 def add(self, model_cls, path, meta, **opts):
     98     # Avoid sending request with empty argument(s)
---> 99     if not(len(path) and len(meta.name) and len(meta.value)):
    100         raise ValueError('Empty value in ' + repr(meta))
    102     resource_type = self._model_class_to_resource_type(model_cls)

TypeError: object of type 'NoneType' has no len()
```

```
irods_path.meta.add(None, None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[30], line 1
----> 1 irods_path.meta.add(None, None)

File ~/git-repos/iBridges/ibridges/meta.py:210, in MetaData.add(self, key, value, units)
    208     raise ValueError("ADD META: Metadata already present")
    209 if self.blacklist:
--> 210     if re.match(self.blacklist, key):
    211         raise ValueError(f"ADD META: Key must not start with {self.blacklist}.")
    212 self.item.metadata.add(key, value, units)

File /usr/local/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/re/__init__.py:167, in match(pattern, string, flags)
    164 def match(pattern, string, flags=0):
    165     """Try to apply the pattern at the start of the string, returning
    166     a Match object, or None if no match was found."""
--> 167     return _compile(pattern, flags).match(string)

TypeError: expected string or bytes-like object, got 'NoneType'
```

I thinksome explicit tests here with a clear message for the user would be much nicer.

Reply:
Let's handle that issue in a separate PR. I think I found those last time as well, but ignored them, because at least the TypeError is the correct error, but we can definitely improve on that.

Reply:
Ok. Then we merge.